File: myFavoriteLocation2/app/src/main/java/com/lq/myfavoritelocation/MapsActivity.java
package com.lq.myfavoritelocation;

import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;
import androidx.room.Database;
import androidx.room.Room;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.material.snackbar.Snackbar;
import com.lq.myfavoritelocation.RoomDatabase.DataBase;
import com.lq.myfavoritelocation.RoomDatabase.FavoriLocation;
import com.lq.myfavoritelocation.RoomDatabase.LocationDAO;
import com.lq.myfavoritelocation.databinding.ActivityMapsBinding;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class MapsActivity extends FragmentActivity implements OnMapReadyCallback, GoogleMap.OnMapLongClickListener {
    ActivityResultLauncher<String> permissionLauncher;
    private GoogleMap mMap;
    private ActivityMapsBinding binding;
    LocationManager manager;
    Location location;
    LatLng lastLatLng;
    DataBase db;
    LocationDAO dao;
    CompositeDisposable disposables = new CompositeDisposable();
    FavoriLocation selectedLocation;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMapsBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setLayout();
        registerLauncher();
        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);
        db =  Room.databaseBuilder(getApplicationContext(), DataBase.class,"FavoriLocation").build();
        dao = db.locationDAO();
    }


    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;
        mMap.setOnMapLongClickListener(this);
        if(ContextCompat.checkSelfPermission(MapsActivity.this, android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED)
        {
            if(ActivityCompat.shouldShowRequestPermissionRationale(MapsActivity.this, android.Manifest.permission.ACCESS_FINE_LOCATION))
            {
                Snackbar.make(binding.getRoot(),"YOU WANT USE MAP YOU SHOULD ALLOW PERMISSION", Snackbar.LENGTH_LONG).setAction("CLICK FOR ALOW PERMISSION ",new View.OnClickListener(){

                    @Override
                    public void onClick(View view) {
                        permissionLauncher.launch(android.Manifest.permission.ACCESS_FINE_LOCATION);
                    }
                }).show(); ;
            }
            else{
                permissionLauncher.launch(android.Manifest.permission.ACCESS_FINE_LOCATION);
            }
        }
        else{
            manager =(LocationManager) getSystemService(Context.LOCATION_SERVICE);

           if(getIntent().getStringExtra("type").equals("oldLocation"))
           {
               FavoriLocation selectLocation =(FavoriLocation) getIntent().getSerializableExtra("selectedLocation");
               lastLatLng = new LatLng(selectLocation.latitude,selectLocation.longitude);
               mMap.addMarker(new MarkerOptions().position(lastLatLng).title(selectLocation.locationName));
               mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(lastLatLng,15));
           }
           else
           {
               location = manager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
               if(location != null)
               {
                   LatLng latLng = new LatLng(location.getLatitude(),location.getLongitude());
                   mMap.addMarker(new MarkerOptions().position(latLng).title("My Last Known Location"));
                   mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng,15));
               }
           }

            manager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, new LocationListener() {
                @Override
                public void onLocationChanged(@NonNull Location location) {

                }
            });
        }
    }

    public void setLayout()
    {
        Intent intent = getIntent();
        String type = intent.getStringExtra("type");
        if(type.equals("newLocation"))
        {
            binding.deleteButton.setVisibility(View.GONE);
        }
        else{
            selectedLocation =(FavoriLocation)intent.getSerializableExtra("selectedLocation");
            System.out.print("selectedLocation.name => "+selectedLocation.locationName);
            binding.locationNameTextView.setText(selectedLocation.locationName);
            binding.addButton.setVisibility(View.GONE);
            binding.locationNameTextView.setEnabled(false);
        }
    }

    public void add(View view)
    {
        this.disposables.add(dao.insert(new FavoriLocation(binding.locationNameTextView.getText().toString(), lastLatLng.longitude, lastLatLng.latitude)).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(this::onCompleted));
    }

    public void delete(View view)
    {
         this.disposables.add(dao.delete(selectedLocation).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(this::onCompleted));
    }

    public void onCompleted()
    {
         Intent intent = new Intent(MapsActivity.this,MainActivity.class);
         startActivity(intent);
         finish();
    }

    public void registerLauncher()
    {
        permissionLauncher = registerForActivityResult(new ActivityResultContracts.RequestPermission(), new ActivityResultCallback<Boolean>() {
            @Override
            public void onActivityResult(Boolean result) {
                if(result)
                {
                    if(ContextCompat.checkSelfPermission(MapsActivity.this,android.Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED)
                    {
                        manager =(LocationManager) getSystemService(Context.LOCATION_SERVICE);
                        location = manager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
                        if(location != null)
                        {

                            LatLng latLng = new LatLng(location.getLatitude(),location.getLongitude());
                            mMap.addMarker(new MarkerOptions().position(latLng).title("My Last Known Location"));
                            mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng,15));

                        }
                        manager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, new LocationListener() {
                            @Override
                            public void onLocationChanged(@NonNull Location location) {

                            }
                        });
                    }
                }
                else{
                    Toast.makeText(MapsActivity.this,"YOU NOT ALLOW MAP PERMISSION", Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    @Override
    public void onMapLongClick(@NonNull LatLng latLng) {
        mMap.clear();
        mMap.addMarker(new MarkerOptions().position(latLng));
        mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng,15));
        lastLatLng = latLng;
    }
}

File: FavouritePlacesWithKotlin/app/src/main/java/com/lq/favouriteplaceswithkotlin/MapsActivity.kt
package com.lq.favouriteplaceswithkotlin

import android.Manifest
import android.app.LocaleManager
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.content.getSystemService
import androidx.core.view.isVisible
import androidx.lifecycle.ReportFragment.Companion.reportFragment
import androidx.room.Room
import androidx.room.Room.databaseBuilder

import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.material.snackbar.Snackbar
import com.lq.favouriteplaceswithkotlin.database.Place
import com.lq.favouriteplaceswithkotlin.database.PlaceDao
import com.lq.favouriteplaceswithkotlin.database.RoomDB
import com.lq.favouriteplaceswithkotlin.databinding.ActivityMapsBinding
import com.lq.favouriteplaceswithkotlin.singlation.SelectPlace
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.schedulers.Schedulers
import kotlinx.coroutines.selects.select
import java.security.Provider

class MapsActivity : AppCompatActivity(), OnMapReadyCallback,GoogleMap.OnMapLongClickListener {

    private lateinit var mMap: GoogleMap
    private lateinit var binding: ActivityMapsBinding
    private lateinit var  permissionLaunch : ActivityResultLauncher<String>;
    private lateinit var manager : LocationManager;
    private lateinit var  listener : LocationListener;
    private  var  lastLoc : Location? = null;
    private lateinit var db : RoomDB;
    private lateinit var dao: PlaceDao;
    var markedLt:Double = 0.0 ;
    var markedLd : Double= 0.0 ;
    private val composite : CompositeDisposable = CompositeDisposable()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMapsBinding.inflate(layoutInflater)
        setContentView(binding.root)
        registerLauncher()
        db = Room.databaseBuilder(
            applicationContext,
            RoomDB::class.java, "Places"
        ).build()
        dao = db.placeDao();
        val mapFragment = supportFragmentManager
            .findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)
    }


    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
        mMap.setOnMapLongClickListener(this)
        val type = intent.getStringExtra("type");
        if(type == "new")
        {
             binding.deleteButton.visibility = View.GONE;
             binding.saveButton.setOnClickListener {


                 composite.add(
                     dao.insertPlace(Place(binding.placeNameEditTextView.text.toString(),markedLt,markedLd))
                         .subscribeOn(Schedulers.io())
                         .observeOn(AndroidSchedulers.mainThread())
                         .subscribe(this::handlerResponse)
                 )
             }
            manager = getSystemService(Context.LOCATION_SERVICE) as LocationManager;
            listener = object:LocationListener{
                override fun onLocationChanged(p0: Location) {

                }

            }
            if(ContextCompat.checkSelfPermission(this,Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED)
            {
                if(ActivityCompat.shouldShowRequestPermissionRationale(this,Manifest.permission.ACCESS_FINE_LOCATION))
                {

                    Snackbar.make(binding.placeNameEditTextView,"YOU WANT USE MAP ,YOU SHOULD ALLOW PERMISSION",Snackbar.LENGTH_INDEFINITE)
                        .setAction("CLICK FOR USE MAP"){
                            permissionLaunch.launch(Manifest.permission.ACCESS_FINE_LOCATION);
                        }
                }
                else{
                    permissionLaunch.launch(Manifest.permission.ACCESS_FINE_LOCATION);
                }
            }
            else{
                lastLoc = manager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
                if(lastLoc != null)
                {
                    val ltld = LatLng(lastLoc!!.latitude,lastLoc!!.longitude);
                    mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(ltld,15f))
                    mMap.addMarker(MarkerOptions().position(ltld).title("LAST LOCATION"))
                    mMap.isMyLocationEnabled =true;
                }
                manager.requestLocationUpdates(LocationManager.GPS_PROVIDER,1000,10f,listener);
            }
        }
        else{
            binding.saveButton.visibility = View.GONE;
            binding.deleteButton.setOnClickListener {
                   composite.add(
                        dao.deletePlace(SelectPlace.selectPlace!!)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(this::handlerResponse)
                   )
            }
            binding.placeNameEditTextView.isEnabled = false;
            val selectPlace = SelectPlace.selectPlace;
             binding.placeNameEditTextView.hint=selectPlace!!.placeName;
             val ltld = LatLng(selectPlace.placeLat,selectPlace.placeLot);
             mMap.addMarker(MarkerOptions().position(ltld).title(selectPlace.placeName))
             mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(ltld,15f));
        }

    }

    override fun onMapLongClick(p0: LatLng) {
        mMap.clear();
        mMap.addMarker(MarkerOptions().position(p0))
        markedLt = p0.latitude;
        markedLd = p0.longitude;
    }

    fun registerLauncher(){
         permissionLaunch=registerForActivityResult(ActivityResultContracts.RequestPermission(),
             ActivityResultCallback { result ->
                   if(result)
                   {
                       if(ContextCompat.checkSelfPermission(this,Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED)
                       {
                           lastLoc = manager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
                           if(lastLoc != null)
                           {
                               val ltld = LatLng(lastLoc!!.latitude,lastLoc!!.longitude);
                               mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(ltld,15f))
                               mMap.addMarker(MarkerOptions().position(ltld).title("LAST LOCATION"))
                               mMap.isMyLocationEnabled =true;
                           }
                           manager.requestLocationUpdates(LocationManager.GPS_PROVIDER,1000,10f,listener);
                       }
                   }
                 else{
                     Toast.makeText(this,"YOU NOT ALLOW PERMISSION",Toast.LENGTH_LONG)
                   }
             })

    }

    fun handlerResponse()
    {
         val intent = Intent(this,MainActivity::class.java);
         intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
         startActivity(intent);
    }

    override fun onDestroy() {
        super.onDestroy()
        composite.clear()
    }
}